crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2418,7 +2421,25 @@ impl<'db> TypeInferenceBuilder<'db> {
                     None
                 };
 
-                bindings_ty(self.db, definitions, unbound_ty)
+                let ty = bindings_ty(self.db, definitions, unbound_ty);
+
+                if ty.may_be_unbound(self.db) && self.unbound_names.insert(name.start()) {
```

Comment:
Add this to the list of things I kinda wish we'd removed Unbound _before_ doing, since it will now need to change to adapt to the removal of Unbound.

On the other hand, adding this diagnostic might also make it easier to remove Unbound, since it means that our tests that assert on unboundness or maybe-unboundness now already have the ability to keep making those assertions, even after Unbound is no longer part of their revealed type. So that's good.